keps/sig-storage/3314-csi-volume-snapshot-delta/README.md

```diff
@@ -0,0 +1,1441 @@
+limiting using constructs found in the `k8s.io/apiserver/pkg/util/flowcontrol`
+package and its sub-packages.
+
+A caching layer to serve subsequent requests with the same input parameters can
+also be considered. Given the complexity associated with caching, more design
+consideration will be needed to determine cache invalidation period, persistence
+mechanism, cache key scheme etc.
+
+#### APIService Resource - CA bundle expiry
+
+The [`APIService` resource][8] provides information on the in-cluster target
+`Service` resource that fronts the aggregated API server. The `spec.caBundle`
+property defines the PEM bundle needed to establish the TLS trust between the
+Kubernetes API server and the service. If expired, the communication between the
+Kubernetes API server and the aggregated API server can be disrupted.
+
+This KEP deems the bundling of certificate management tools with the CBT
+components to be out-of-scope. User should incorprate the management of this CA
+bundle into their overall cluster PKI strategy.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The proposed design involves extending CSI with the `VolumeSnapshotDelta` and
+`DriverDiscovery` custom resources. Storage providers can opt in to support this
+feature by implementing the `LIST_BLOCK_SNAPSHOT_DELTAS` capability in their
+CSI drivers.
+
+The `VolumeSnapshotDelta` resource is a namespace-scoped resource. It must be
+created in the same namespace as the base and target CSI `VolumeSnapshot`s. On
+the other hand, the `DriverDiscovery` resource will be implemented as a
+cluster-scoped resource.
+
+### CBT Datapath Worklow
+
+The CSI CBT is made up of three components:
+
+* An aggregated API server to serve CBT requests, initiated by the creation of
+`VolumeSnapshotDelta` resources
+* A CRD controller that watches and reconciles `DriverDiscovery` resources
+* A [CSI CBT sidecar][9] that makes a CSI driver discoverable by the aggregated
+API server
+
+![CBT Step 1](./img/cbt-step-01.png)
+
+When CSI CBT is deployed on a Kubernetes cluster, the `cbt-aggapi` aggregated
+API server registers itself with the `kube-aggregator` to claim the URL path
+of the `v1alpha1.cbt.storage.k8s.io` group version, as defined in the
+`APIService` resource.
+
+The `driver-discovery` CRD controller starts a reconciler to watch and
+reconcile `DriverDiscovery` resources.
+
+A storage provider that supports CBT functionalities needs to embeds the
+`cbt-http` sidecar in its CSI driver. As the CSI driver initializes, the
+`cbt-http` sidecar publishes itself by creating a `DriverDiscovery` resource
+with information on where the CSI driver's `cbt-http` endpoint is.
```

Comment:
How the sidecar knows name of the Service that should be in `DriverDiscovery`? How does it know the CA bundle?

IMO the object should be created by an install tool (script, helm chart, operator, kustomize, ...) and not the sidecar.

Reply:
The [prototype][1] initialized `DriverDiscovery` in the sidecar to ensure that the resource is created when the sidecar is deployed. But yes, it makes sense to move this to an install tool, especially if we want to make this into a more generic service configuration, as discussed above.

[1]: https://github.com/ihcsim/cbt-aggapi/blob/7631d13050c71bb2c3b810fce4c46ba3f7f9fc7d/cmd/mock/http/main.go#L68-L114